src/profiler.ts

```diff
@@ -13,57 +13,203 @@
+const pjson = require('../../package.json');
+const API = 'https://cloudprofiler.googleapis.com/v2';
 
-export interface ProfilerConfig extends AuthenticationConfig {
-  projectId: string;
-  logLevel: number;
-  serviceContext: {service: string; version?: string;};
-  instance: string;
-  zone: string;
-  disableTime: boolean;
-  disableHeap: boolean;
+// Interface for body of response from Stackdriver Profiler API when creating
+// profile and used as body of request to Stackdriver Profiler API when
+// uploading a profile.
+// Public for testing.
+export interface RequestProfile {
+  name?: string;
+  profileType?: string;
+  duration?: any;
+  profileBytes?: string;
+  labels?: {instance?: string; zone?: string};
 }
 
-// TODO: finish implementing Profiler.
-// TODO: add stop() method to stop profiling.
+// Converts a profile to a compressed, base64 encoded string.
+function profileBytes(p: perftools.profiles.IProfile): Promise<string> {
+  let time = new Date((p.timeNanos as number) / 1000000);
+  let pwriter = perftools.profiles.Profile.encode(p);
+  let buffer = new Buffer(pwriter.finish());
+  return new Promise<string>((resolve, reject) => {
+    zlib.gzip(buffer, (err: Error, gzBuf: Buffer) => {
+      err ? reject(err) : resolve(gzBuf.toString('base64'));
+    });
+  });
+}
+
+// Service object for communicating with Stackdriver Profiler API.
+class ProfilerServiceObject extends common.ServiceObject {
+  constructor(options: AuthenticationConfig) {
+    options = common.util.normalizeArguments(null, options);
+    const config = {
+      baseUrl: API,
+      scopes: ['https://www.googleapis.com/auth/monitoring.write'],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(config, options), baseUrl: '/'});
+  }
+}
+
+// Polls Stackdriver Profiler server for instructions on behalf of a task and
+// collects and uploads profiles as requested.
 export class Profiler {
-  config: ProfilerConfig;
-  logger: Logger;
-  service: Service;
+  private config: ProfilerConfig;
+  private logger: Logger;
+  private serviceObject: ServiceObject;
+  private profileTypes: string[];
+  private continueProfiling: boolean;
+
+  // Public for testing.
+  timeProfiler: TimeProfiler|undefined;
 
   constructor(config: ProfilerConfig) {
     this.config = config;
-
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
-
+    this.serviceObject = new ProfilerServiceObject(config);
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    // TODO: enable heap profiling once heap-profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push('WALL');
+      this.timeProfiler = new TimeProfiler(1000);
```

Comment:
Declare a constant variable at the top or in the class instead of providing it here.

Secondly, should this be configurable by the user and exposed in the config instead?

Reply:
I think we need to decide what characteristics of profiling should be configurable and then make those things configurable.
I'll file an issue, and that can be a separate PR going forward.